au/overflow_boundary_test.cc

```diff
@@ -0,0 +1,285 @@
+                Eq(MinGood<StaticCast<int64_t, int32_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int64_t, uint32_t>, ImplicitLimits<uint32_t>>::value()),
+                Eq(MinGood<StaticCast<int64_t, uint32_t>>::value()));
+}
+
+TEST(StaticCast, MinGoodUnchangedWithExplicitLimitLessConstrainingThanExistingResult) {
+    // In these cases, we are applying a non-trivial lower limit (i.e., it is higher than the
+    // `lowest()` value), but it does not constrain the result enough to change it.
+
+    struct DoubleLimitTwiceFloatLowest : NoUpperLimit<double> {
+        static constexpr double lower() {
+            return static_cast<double>(std::numeric_limits<float>::lowest()) * 2.0;
+        }
+    };
+
+    EXPECT_THAT((MinGood<StaticCast<float, double>, DoubleLimitTwiceFloatLowest>::value()),
+                Eq(MinGood<StaticCast<float, double>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int32_t, double>, DoubleLimitTwiceFloatLowest>::value()),
+                Eq(MinGood<StaticCast<int32_t, double>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<uint16_t, double>, DoubleLimitTwiceFloatLowest>::value()),
+                Eq(MinGood<StaticCast<uint16_t, double>>::value()));
+
+    struct FloatLimitHalfFloatLowest : NoUpperLimit<float> {
+        static constexpr float lower() { return std::numeric_limits<float>::lowest() / 2.0f; }
+    };
+
+    EXPECT_THAT((MinGood<StaticCast<uint64_t, float>, FloatLimitHalfFloatLowest>::value()),
+                Eq(MinGood<StaticCast<uint64_t, float>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int64_t, float>, FloatLimitHalfFloatLowest>::value()),
+                Eq(MinGood<StaticCast<int64_t, float>>::value()));
+
+    struct SignedLimitHalfInt64Lowest : NoUpperLimit<int64_t> {
+        static constexpr int64_t lower() { return std::numeric_limits<int64_t>::lowest() / 2; }
+    };
+
+    EXPECT_THAT((MinGood<StaticCast<uint32_t, int64_t>, SignedLimitHalfInt64Lowest>::value()),
+                Eq(MinGood<StaticCast<uint32_t, int64_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int32_t, int64_t>, SignedLimitHalfInt64Lowest>::value()),
+                Eq(MinGood<StaticCast<int32_t, int64_t>>::value()));
+}
+
+TEST(StaticCast, MinGoodUnchangedForUnsignedDestinationAndExplicitLimitOfZero) {
+    EXPECT_THAT((MinGood<StaticCast<uint8_t, uint16_t>, LowerLimitIsZero<uint16_t>>::value()),
+                Eq(MinGood<StaticCast<uint8_t, uint16_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<int32_t, uint64_t>, LowerLimitIsZero<uint64_t>>::value()),
+                Eq(MinGood<StaticCast<int32_t, uint64_t>>::value()));
+
+    EXPECT_THAT((MinGood<StaticCast<double, uint32_t>, LowerLimitIsZero<uint32_t>>::value()),
+                Eq(MinGood<StaticCast<double, uint32_t>>::value()));
+}
+
+TEST(StaticCast, MinGoodCappedByExplicitFloatLimit) {
+    struct FloatLowerLimitMinusOne : NoUpperLimit<float> {
+        static constexpr float lower() { return -1.0f; }
```

Comment:
I suppose here as well, we could just use `-1`?

```suggestion
        static constexpr float lower() { return -1; }
```